### ISSUE TITLE
Vernemq crash and 100% CPU consumption when integrated with mongodb

### ISSUE THREAD

**Issue by KnolBee** [erlio/vernemq#380]
### Environment

- VerneMQ Version:1.0.1
- OS:Ubuntu 16.04
- Erlang/OTP version (if building from source):

### Expected behavior

VerneMQ shouldn't stall or crash when integrated with MongoDB.

### Actual behaviour

- VerneMQ stalls while consuming 100% cpu cylces.

Setup/Steps to reproduce:

1.   AWS setup
2.  1 Load balancer running haproxy, configured to handle millions of sockets: m4.4xlarge 64 GB instance
3.  3x Vernemq cluster : t2.2xlarge 32GB instance
4.  1x Mongodb instance : m4.2xlarge
5.  Configure vernemq to authenticate from mongodb
6.  Create millions of entries in mongodb vmq_acl_auth database for testing. sample looks like 
{"mountpoint" : "", "client_id" : "0", "username" : "iuconnector", "passhash" : "$2a$06$nV2TRXCc/1AJYVGrAzuO7.ODHzF7jEXdqPLM6oziPif5cOM2rm8wO", "publish_acl" : [ { "pattern" : "#" } ], "subscribe_acl" : [ { "pattern" : "#" } ] }
  ids range from 0-25500000, totally 25 million entries in mongodb instance . Make sure you have created index on this table so we do not hit query bottleneck.
7. Nodejs app to create 50k connections. Start with 1, scale to 16 gradually, everything works
8. Now stop all nodejs app, start 6 apps simultaneously, one app per one aws instance. After a while VerneMQ goes bad
9. Stop load balancer so no request reaches vernemq, configurm using netstat command.  But vernemq continues to consume 100% cpu, nothing shows up in log, crash report has few entries. MongoDB flooded with queries for vmq_acl_auth database even after all connections to vernemq is blocked by stopping the load balancer.

Attached zip has

1. Image showing 6 load test instance in aws
2. Image showing 100% cpu consumption by vernemq
3. Image showing there are no connections for vernemq at the time when it went bad
4. node js app i have used to test
5. logs of vernemq

My Observation:

Scaling gradually works, i.e I can take connections upto 1 million if i run the load test app one by one, but if i start test even 5 instance at once, vernemq hangs.

Thanks and Regards,
Suresh
[crashreport.zip](https://github.com/erlio/vernemq/files/996893/crashreport.zip)

**Comment by ioolkos** [comment]
Thanks @KnolBee for reporting.

Whoa hell unleashed! This might be a version of
https://github.com/erlio/vernemq/issues/341
where the root cause would be a timeout in the `auth_on_register` when called over vmq_diversity (Lua)
@dergraf I think we have now multiple scenarios/reports pointing to this, so a VerneMQ side issue is quite probable (?)

**Comment by KnolBee** [comment]
@ioolkos I could get a similar error with single load test system by changing the test script to send high connect request rate. When i rate limit in HAProxy as "rate-limit sessions 10", the error is mitigated, but this is too low value for rate-limit. It's something to do with a burst of connect request to Verne MQ. 

usage would be like: node index.js <vernemqipaddress>

[loadtestjs_crash.zip](https://github.com/erlio/vernemq/files/998554/loadtestjs_crash.zip)

**Comment by ioolkos** [comment]
Controlling setup rates should be part of all load testing tools. If you just hit a TCP endpoint with 50k connection requests in a for loop, I think all bets are off anyway to where the contention lies. 

If you plan for 1k connections per second setup rate, configure 1k acceptors (`listener.nr_of_acceptors`) as a start.
The next step is to think about register rate if you do an auth plugin, that is what rates `auth_on_register` has to provide (and test if it really does).
The third step is to think about subscribe rates, which might overload a VerneMQ cluster etc.

Also think about the small stuff. Are you're clients re-connecting immediately? try without etc.

I hope you (and us) find a test setup that works for you, it's great to see your efforts! :)

Thanks for your node script, i'll try to test it locally on my machine.

**Comment by KnolBee** [comment]
thanks for the inputs. I did try setting nr_of_acceptors to 1000(tried 10000 too), but ending with same issue even when i control the new connection rate 50 at HAProxy level, do you have any metrics on how much load in terms of connection rate  vernemq can handle for the t2.2xlarge instance, i will also try higher configuration with more cpu/ram to see if it helps.

**Comment by dergraf** [comment]
#381 Should fix the crashes.

**Comment by dergraf** [comment]
Setting `vmq_diversity.mongodb.pool_size` to a higher value then the default (`5`) may help too.

**Comment by KnolBee** [comment]
setting vmq_diversity.mongodb.pool_size to 100 and 1000 didn't help, Trying to build with the #381 fix. When can this be part of the official release?

**Comment by larshesel** [comment]
#381 is now merged to master and will be part of the 1.0.2 release. When we'll release that isn't yet decided.

**Comment by KnolBee** [comment]
@larshesel Finally could test this fix, it works, thanks a lot, you guys doing great work.

**Comment by KnolBee** [comment]
Just to understand the problem, i get these logs now,

2017-05-19 06:38:41.459 [error] <0.18942.1>@vmq_diversity_script:call_function:69 can't call into Lua sandbox for function auth_on_register due to {timeout,{gen_server,call,[<0.317.0>,{call_function,auth_on_register,[{addr,<<"10.0.0.119">>},{port,38095},{mountpoint,<<>>},{client_id,<<"11928131">>},{username,<<"testusername">>},{password,<<"testpassword">>},{clean_session,true}]}]}}

Does this mean MongoDB takes time? I have configured MongoDB to have an index for the column client_id, so its unlikely its the bottleneck, can we increase the timeout using any config?

**Comment by dergraf** [comment]
Closing this in favour of https://github.com/erlio/vernemq/issues/389 as the cause of the issue is the same, explained in https://github.com/erlio/vernemq/issues/389#issuecomment-308674406
